File: app/src/main/java/org/dhis2/utils/CodeGeneratorImpl.java
package org.dhis2.utils;

import androidx.annotation.NonNull;

import java.util.Random;

/**
 * Created by ppajuelo on 09/01/2018.
 *
 */

public class CodeGeneratorImpl implements CodeGenerator {
    private static final String LETTERS = "abcdefghijklmnopqrstuvwxyz"
            + "ABCDEFGHIJKLMNOPQRSTUVWXYZ";

    private static final String ALLOWED_CHARS = "0123456789" + LETTERS;
    private static final int NUMBER_OF_CODEPOINTS = ALLOWED_CHARS.length();
    private static final int CODESIZE = 11;

    private Random sr = new Random();

    @NonNull
    @Override
    public String generate() {
        char[] randomChars = new char[CODESIZE];

        // First char should be a letter
        randomChars[0] = LETTERS.charAt(sr.nextInt(LETTERS.length()));

        for (int i = 1; i < CODESIZE; ++i) {
            randomChars[i] = ALLOWED_CHARS.charAt(
                    sr.nextInt(NUMBER_OF_CODEPOINTS));
        }

        return new String(randomChars);
    }
}

File: app/src/main/java/org/dhis2/utils/CodeGeneratorImpl.kt
package org.dhis2.utils

import java.util.*

class CodeGeneratorImpl: CodeGenerator {
    val LETTERS = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
    val ALLOWED_CHARS = "0123456789$LETTERS"
    val NUMBER_OF_CODEPOINTS = ALLOWED_CHARS.length
    val CODESIZE = 11
    val random = Random()

    override fun generate(): String {
        val randomChars = CharArray(CODESIZE)
        // First char should be a letter
        randomChars[0] = LETTERS[random.nextInt(LETTERS.length)]
        for (x in 1..CODESIZE) {
            randomChars[x] = ALLOWED_CHARS[random.nextInt(NUMBER_OF_CODEPOINTS)]
        }
        return randomChars.toString()
    }

}